src/clj_commons/byte_streams.clj

```diff
@@ -576,35 +577,51 @@
 ;; file => readable-channel
 (def-conversion ^{:cost 0} [File ReadableByteChannel]
   [file]
-  (.getChannel (FileInputStream. file)))
+  (-> file
+      (.toPath)
+      (FileChannel/open (into-array StandardOpenOption
+                                    [StandardOpenOption/READ]))))
 
 ;; file => writable-channel
 (def-conversion ^{:cost 0} [File WritableByteChannel]
   [file {:keys [append?] :or {append? true}}]
-  (.getChannel (FileOutputStream. file (boolean append?))))
+  (let [option-array (into-array StandardOpenOption
+                                 (cond-> [StandardOpenOption/CREATE
+                                          StandardOpenOption/WRITE]
+                                         append?
+                                         (conj StandardOpenOption/APPEND)))]
+    (-> file
+        (.toPath)
+        (FileChannel/open option-array))))
 
 (def-conversion ^{:cost 0} [File (seq-of ByteBuffer)]
   [file {:keys [chunk-size writable?] :or {chunk-size (int 2e9), writable? false}}]
   (let [^RandomAccessFile raf (RandomAccessFile. file (if writable? "rw" "r"))
         ^FileChannel fc (.getChannel raf)
+        close-fn (fn []
+                   (.close raf)
+                   (.close fc))
         buf-seq (fn buf-seq [offset]
-                  (when-not (<= (.size fc) offset)
-                    (let [remaining (- (.size fc) offset)]
+                  (when (and (.isOpen fc)
+                             (> (.size fc) offset))
+                    (let [remaining (- (.size fc) offset)
+                          close-after-reading? (<= remaining chunk-size)]
                       (lazy-seq
                         (cons
-                          (.map fc
-                            (if writable?
-                              FileChannel$MapMode/READ_WRITE
-                              FileChannel$MapMode/READ_ONLY)
-                            offset
-                            (min remaining chunk-size))
+                          (let [mbb (.map fc
+                                          (if writable?
+                                            FileChannel$MapMode/READ_WRITE
+                                            FileChannel$MapMode/READ_ONLY)
+                                          offset
+                                          (min remaining chunk-size))]
+                            (when close-after-reading?
+                              (close-fn))
```

Comment:
> It's been updated to automatically close the file upon exhaustion (Maybe this should be an option? We wouldn't want to stop tailing a growing log file just because we temporarily caught up to the end...)

To me, as we open the resource, we have the responsibility to close it.

Reply:
I agree, but that comment's not about responsibility, but default behavior. For a static file, we should close upon reading the last byte, since with the `File -> (seq-of ByteBuffer)` conversion, there's no outside resource being supplied that the user knows they have to close. Otherwise, they have to close the seq when done with it (which isn't so bad, if it came down to it, but it's not intuitive.)

The catch is, what if you're reading a file that's growing, like tailing a log file? You effectively *want* an infinite seq. As it is now, if the lazy-seq ever catches up to the current end of the file, it'll stop and close, even if the file is still growing.

The more I think about it, the more I think there should be an option to turn off auto-closing if you don't want it.

Reply:
Hrrm. Unfortunately, a lot more of the code in that conversion assumes the file is static. In particular, `(> (.size fc) offset)` will end the seq, even if we don't `.close` the channel. If we catch up to the end of the file, it'll return nil and the lazy-seq stops, when we would prefer it block if the file is growing.

That test also means shrinking or truncating the file will make it fail to .close, though to be honest, a seq across a shrinking file doesn't make logical sense anyway.

I think I'll just add a comment for now. This will require reworking to add blocking/async behavior if we wanted it to handle growing files.